### PR TITLE
Fix analysis over umbrella apps

### LINKED ISSUE #178: Analysis not working over umbrella apps
As per https://github.com/project-fifo/rebar3_lint/issues/26#issue-826719550.

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
Closes #178.

- [x] make sure tests are passing in Windows
- [x] make sure to add tests for updated functionality (namely we should test folder structure as per umbrella apps)

---
**Note**: this includes a small patch to run the tests on Windows.

### CODE REVIEW

File: src/elvis_file.erl [hunk 1]
```diff
@@ -86,37 +86,42 @@ find_files(Dirs, Pattern) ->
...
+                                   DirOrFile =:= Path
+                               end,
+                               ExpandedFilter)
+                 end,
+                 Files).
+
 %% @doc Filter files based on the glob provided.
 -spec filter_files([file()], [string()], string(), [string()]) -> [file()].
 filter_files(Files, Dirs, Filter, IgnoreList) ->
-    AppendFilter = fun(Dir) ->
```
Comment (paulo-ferraz-oliveira) [inline src/elvis_file.erl:92]: There's a fix here, and a bit of refactoring.

The refactoring is basically to try and improve readability (no more functions declared as variables, for example).

The fix is to have analysis work over umbrella projects, where input `dirs` might be `apps/**/src`.

Considerations on the algo:
1. we still create a list of filters (`dir`+`filter`) that we expand with `wildcard`, (`ExpandedFilters`)
2. we go over a list of previously found files to check if they "fall" in the filters, (`Found`)
3. we remove the `ignore` element from the previous list (function result).

We do this without resorting to regex, which was not only confusing, as-was, but error prone (`**` not accepted).

Now, the code for `Found` seems oddly similar to that of function `find_files/2` further above, so some more refactoring might be in order.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -29,3 +30,29 @@ jobs:
       - run: |
           rebar3 escriptize
           _build/default/bin/elvis_core
+  ci-windows:
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ci.yml:33]: Let's hope I can get this working without many issues

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -9,15 +9,16 @@ on:
...
   ci:
-    name: Run checks and tests over ${{matrix.otp_vsn}} and ${{matrix.os}}
+    name: >
+      Run Linux-based checks and tests over ${{matrix.otp_vsn}} and ${{matrix.os}}
     runs-on: ${{matrix.os}}
     strategy:
       matrix:
         otp_vsn: [19.3.6.13, 20.3.8.26, 21.3.8.17, 22.3.4.9, 23.1.5]
-        os: [ubuntu-latest]
+        os: [ubuntu-18.04]
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ci.yml:18]: This break due to wx -related stuff being renamed.

**Comment by elbrujohalcon** [comment]
I like it… let's see if the windows-ci starts working at some point.

File: src/elvis_file.erl [hunk 4]
```diff
@@ -86,37 +86,46 @@ find_files(Dirs, Pattern) ->
...
+    Filter;
+dir_to(Filter, Dir) ->
+    filename:join(Dir, Filter).
+
+file_in(ExpandedFilter, Files) ->
+    lists:filter(fun (_File = #{ path := Path }) ->
+                     lists:any(fun (DirOrFile) ->
+                                   DirOrFile =:= Path
+                               end,
+                               ExpandedFilter)
```
[the remark this answers is not in the document]
Comment (paulo-ferraz-oliveira) [reply src/elvis_file.erl:99]: It was soooo late in the night. (also, I don't know why I was under the impression `lists:member` did a full list scan, which would make no sense, but still...).

**Comment by paulo-ferraz-oliveira** [comment]
@elbrujohalcon, I moved the tasks from the issue here (check the description) so there's visibility. I still want to had tests lest we end up in the same situation.

**Comment by paulo-ferraz-oliveira** [comment]
@elbrujohalcon: pushed some tests. I'm not sure we need something more elaborate. It walks an expected (and advertised, in `rebar3_lint` folder structure) using regex.
Should we probably throw in a `ignore` test?

**Comment by paulo-ferraz-oliveira** [comment]
I segregated the Windows CI thing to an Issue, lest we're blocked because of it. I'd wrongly assumed stuff was already working and would have only to deal with the container stuff, but it seems not.

**Comment by paulo-ferraz-oliveira** [comment]
@joaohf, do you want to try these changes out to see if they fix your specific issue?

**Comment by joaohf** [comment]
@paulo-ferraz-oliveira sure, I'll try today.

**Review by elbrujohalcon — APPROVED** [review]
I like how this looks.

**Comment by joaohf** [comment]
Alright, the fix works for me. Really good. Thank you guys.

**Comment by paulo-ferraz-oliveira** [comment]
@joaohf, is it acceptable to rely on `master` for the time being, before we cut a release? I don't expect this to break for anyone relying on `master`, but the next few days will tell 😄

**Comment by joaohf** [comment]
Sure, don't worry. I can rely on master until there.

**Pull request merged**